database/dbutil.go

```diff
@@ -0,0 +1,286 @@
+// Copyright 2018 clair authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"github.com/deckarep/golang-set"
+)
+
+// DeduplicateNamespaces deduplicates a list of namespaces.
+func DeduplicateNamespaces(namespaces ...Namespace) []Namespace {
+	nsSet := mapset.NewSet()
+	for _, ns := range namespaces {
+		nsSet.Add(ns)
+	}
+
+	result := make([]Namespace, 0, nsSet.Cardinality())
+	for ns := range nsSet.Iter() {
+		result = append(result, ns.(Namespace))
+	}
+
+	return result
+}
+
+// DeduplicateFeatures deduplicates a list of list of features.
+func DeduplicateFeatures(features ...Feature) []Feature {
+	fSet := mapset.NewSet()
+	for _, f := range features {
+		fSet.Add(f)
+	}
+
+	result := make([]Feature, 0, fSet.Cardinality())
+	for f := range fSet.Iter() {
+		result = append(result, f.(Feature))
+	}
+
+	return result
+}
+
+// PersistPartialLayer wraps session PersistLayer function with begin and
+// commit.
+func PersistPartialLayer(datastore Datastore, layer *Layer) error {
+	tx, err := datastore.Begin()
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	if err := tx.PersistLayer(layer.Hash, layer.Features, layer.Namespaces, layer.By); err != nil {
+		return err
+	}
+
+	return tx.Commit()
+}
+
+// PersistFeatures wraps session PersistFeatures function with begin and commit.
+func PersistFeatures(datastore Datastore, features []Feature) error {
+	tx, err := datastore.Begin()
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	if err := tx.PersistFeatures(features); err != nil {
+		return err
+	}
+	return tx.Commit()
+}
+
+// PersistNamespaces wraps session PersistNamespaces function with begin and
+// commit.
+func PersistNamespaces(datastore Datastore, namespaces []Namespace) error {
```

Comment:
We should come up with a naming convention for the functions that are just wrapping calls in a transaction.